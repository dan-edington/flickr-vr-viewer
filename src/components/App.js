import React, { Component } from 'react';
import { Entity, Scene } from 'aframe-react';
import aframe from 'aframe';
import Camera from './Camera';
import Skybox from './Skybox';
import Search from './Search';
import SearchOpener from './SearchOpener';

class App extends Component {

  constructor() {

    super();
  
    this.state = {
      imageURL: null,
      searchIsVisible: false,
      searchPosition: { x: 0, y: 0, z: 0 },
      searchRotation: 0
    }

    this.doSearch = this.doSearch.bind(this);
    this.determineSearchPositionAndVisibility = this.determineSearchPositionAndVisibility.bind(this);

  }

  componentDidMount() {

    this.doSearch('Japan');

  }

  doSearch(searchTerm) {

    const url = `https://api.flickr.com/services/rest/?method=flickr.photos.search&api_key=8c226f8233af4f331fc8c0322631bfdb&text=${searchTerm}&extras=url_k&group_id=44671723%40N00&format=json&nojsoncallback=1`;

    fetch(url)
      .then(res => res.json())
      .then(res => {
        const photoURLs = res.photos.photo.map(photo => {
          return photo.url_k;
        }).filter(photo => {
          return !!photo;
        });

        this.setState({
          imageURL: photoURLs[Math.floor(Math.random() * (photoURLs.length - 1))]
        });

      });   

  }

  determineSearchPositionAndVisibility(visibility) {

    let searchXPosition = this.state.searchPosition.x;
    let searchZPosition = this.state.searchPosition.z;
    let cameraRotationY = this.state.searchRotation;

    if(visibility && this.refs.camera) {

      const cameraEntity = this.refs.camera.refs.cameraEntity.el;
      cameraRotationY = cameraEntity.getAttribute('rotation').y;
      searchZPosition = Math.cos(cameraRotationY * Math.PI / 180) * -5;
      searchXPosition = Math.sin(cameraRotationY * Math.PI / 180) * -5;

    }

    this.setState({
      searchIsVisible: visibility,
      searchPosition: {
        x: searchXPosition,
        y: 0,
        z: searchZPosition
      },
      searchRotation: cameraRotationY
    });

  }

  renderSearch() {

    if(this.state.searchIsVisible && this.refs.camera) {

      return(
        <Search
          position={ this.state.searchPosition }
          rotation={{ x: 0, y: this.state.searchRotation, z: 0 }}
          determineSearchPositionAndVisibility={this.determineSearchPositionAndVisibility}
          doSearch={this.doSearch} />
      );

    }

    return false;

  }

  render() {
    return(
      <Scene>
        <Skybox imageURL={this.state.imageURL} />
        <Camera fuseTime="750" ref="camera" />
        { this.renderSearch() }
        <SearchOpener determineSearchPositionAndVisibility={this.determineSearchPositionAndVisibility} />
      </Scene>
    )
  }

};

export default App;