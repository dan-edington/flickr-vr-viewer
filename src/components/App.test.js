import aframe from 'aframe';
import React from 'react';
import { shallow, mount } from 'enzyme';
import fetchMock from 'fetch-mock';
import App from './App';

describe('App', () => {

  let app;
  const searchQuery = 'Paris';
  const url = `https://api.flickr.com/services/rest/?method=flickr.photos.search&api_key=8c226f8233af4f331fc8c0322631bfdb&text=${searchQuery}&extras=url_k&group_id=44671723%40N00&format=json&nojsoncallback=1`;

  const mockResponse = {
    photos: {
      photo: [{
        farm: 5,
        id: "25527609148",
        isfamily: 0,
        isfriend: 0,
        ispublic: 1,
        owner: "146123903@N04",
        secret: "bffd3b0efa",
        server: "4682",
        title: "Las Vegas (Paradise)",
        url_k: "https://farm5.staticflickr.com/4682/25527609148_bffd3b0efa_k.jpg"
      },
      {
        farm: 5,
        id: "37181315310",
        isfamily: 0,
        isfriend: 0,
        ispublic: 1,
        owner: "66252312@N00",
        secret: "8f57a632dd",
        server: "4509",
        title: "The Catacombs of Paris (360ยบ Photo)",
        url_k: "https://farm5.staticflickr.com/4509/37181315310_8f57a632dd_k.jpg"
      },
      {
        farm: 5,
        id: "37181315310",
        isfamily: 0,
        isfriend: 0,
        ispublic: 1,
        owner: "66252312@N00",
        secret: "8f57a632dd",
        server: "4509",
        title: "The Catacombs of Paris (360ยบ Photo)",
        url_k: undefined
      },
      {
        farm: 5,
        id: "36723254644",
        isfamily: 0,
        isfriend: 0,
        ispublic: 1,
        owner: "66252312@N00",
        secret: "e0016b2158",
        server: "4395",
        title: "The Louvre (360ยบ Photo)",
        url_k: "https://farm5.staticflickr.com/4395/36723254644_e0016b2158_k.jpg"
      }
      ]
    }
  }

  const expectedResult = [
    'https://farm5.staticflickr.com/4682/25527609148_bffd3b0efa_k.jpg',
    'https://farm5.staticflickr.com/4509/37181315310_8f57a632dd_k.jpg',
    'https://farm5.staticflickr.com/4395/36723254644_e0016b2158_k.jpg'
  ];

  fetchMock.get(url, mockResponse);

  beforeEach(() => {
    app = shallow(<App />);
  });

  it('renders correctly', () => {

    expect(app).toMatchSnapshot();

  });

  it('has one <a-sky> element', () => {

    expect(app.find('a-sky')).toHaveLength(1);

  });

  describe('When the doSearch() method is called', () => {

    it('returns an array of photo urls', async () => {

      const searchResult = await app.instance().doSearch(searchQuery);
      expect(searchResult).toEqual(expect.arrayContaining(expectedResult));

    });

  });

  describe('When the loadImage() method is called', () => {

    it('sets the src attribute of the a-sky element to the first value in the photo url array', () => {

      app.instance().loadImage(expectedResult[0]);
      expect(app.find('a-sky').props().src).toEqual(expectedResult[0]);

    });

  });

});