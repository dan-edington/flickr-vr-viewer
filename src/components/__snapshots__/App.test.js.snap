// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <a-sky />,
        <a-entity
          position="0 0 0"
        >
          <a-entity
            camera={true}
            look-controls={true}
          >
            <a-entity
              cursor="maxDistance: 30; fuse: true"
              geometry="primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;"
              material="color: black; shader: flat"
              position="0 0 -3"
            >
              <a-animation
                attribute="scale"
                begin="fusing"
                dur="1500"
                easing="ease-in"
                fill="backwards"
                from="1 1 1"
                to="0.2 0.2 0.2"
              />
            </a-entity>
          </a-entity>
        </a-entity>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": "sky",
        "rendered": null,
        "type": "a-sky",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <a-entity
            camera={true}
            look-controls={true}
          >
            <a-entity
              cursor="maxDistance: 30; fuse: true"
              geometry="primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;"
              material="color: black; shader: flat"
              position="0 0 -3"
            >
              <a-animation
                attribute="scale"
                begin="fusing"
                dur="1500"
                easing="ease-in"
                fill="backwards"
                from="1 1 1"
                to="0.2 0.2 0.2"
              />
            </a-entity>
          </a-entity>,
          "position": "0 0 0",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "camera": true,
            "children": <a-entity
              cursor="maxDistance: 30; fuse: true"
              geometry="primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;"
              material="color: black; shader: flat"
              position="0 0 -3"
            >
              <a-animation
                attribute="scale"
                begin="fusing"
                dur="1500"
                easing="ease-in"
                fill="backwards"
                from="1 1 1"
                to="0.2 0.2 0.2"
              />
            </a-entity>,
            "look-controls": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <a-animation
                attribute="scale"
                begin="fusing"
                dur="1500"
                easing="ease-in"
                fill="backwards"
                from="1 1 1"
                to="0.2 0.2 0.2"
              />,
              "cursor": "maxDistance: 30; fuse: true",
              "geometry": "primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;",
              "material": "color: black; shader: flat",
              "position": "0 0 -3",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "attribute": "scale",
                "begin": "fusing",
                "dur": "1500",
                "easing": "ease-in",
                "fill": "backwards",
                "from": "1 1 1",
                "to": "0.2 0.2 0.2",
              },
              "ref": null,
              "rendered": null,
              "type": "a-animation",
            },
            "type": "a-entity",
          },
          "type": "a-entity",
        },
        "type": "a-entity",
      },
    ],
    "type": "a-scene",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <a-sky />,
          <a-entity
            position="0 0 0"
          >
            <a-entity
              camera={true}
              look-controls={true}
            >
              <a-entity
                cursor="maxDistance: 30; fuse: true"
                geometry="primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;"
                material="color: black; shader: flat"
                position="0 0 -3"
              >
                <a-animation
                  attribute="scale"
                  begin="fusing"
                  dur="1500"
                  easing="ease-in"
                  fill="backwards"
                  from="1 1 1"
                  to="0.2 0.2 0.2"
                />
              </a-entity>
            </a-entity>
          </a-entity>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": "sky",
          "rendered": null,
          "type": "a-sky",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <a-entity
              camera={true}
              look-controls={true}
            >
              <a-entity
                cursor="maxDistance: 30; fuse: true"
                geometry="primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;"
                material="color: black; shader: flat"
                position="0 0 -3"
              >
                <a-animation
                  attribute="scale"
                  begin="fusing"
                  dur="1500"
                  easing="ease-in"
                  fill="backwards"
                  from="1 1 1"
                  to="0.2 0.2 0.2"
                />
              </a-entity>
            </a-entity>,
            "position": "0 0 0",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "camera": true,
              "children": <a-entity
                cursor="maxDistance: 30; fuse: true"
                geometry="primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;"
                material="color: black; shader: flat"
                position="0 0 -3"
              >
                <a-animation
                  attribute="scale"
                  begin="fusing"
                  dur="1500"
                  easing="ease-in"
                  fill="backwards"
                  from="1 1 1"
                  to="0.2 0.2 0.2"
                />
              </a-entity>,
              "look-controls": true,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <a-animation
                  attribute="scale"
                  begin="fusing"
                  dur="1500"
                  easing="ease-in"
                  fill="backwards"
                  from="1 1 1"
                  to="0.2 0.2 0.2"
                />,
                "cursor": "maxDistance: 30; fuse: true",
                "geometry": "primitive: ring; radiusInner: 0.05; radiusOuter: 0.09;",
                "material": "color: black; shader: flat",
                "position": "0 0 -3",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "attribute": "scale",
                  "begin": "fusing",
                  "dur": "1500",
                  "easing": "ease-in",
                  "fill": "backwards",
                  "from": "1 1 1",
                  "to": "0.2 0.2 0.2",
                },
                "ref": null,
                "rendered": null,
                "type": "a-animation",
              },
              "type": "a-entity",
            },
            "type": "a-entity",
          },
          "type": "a-entity",
        },
      ],
      "type": "a-scene",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
